Cython/Compiler/Nodes.py

```diff
@@ -2072,26 +2071,28 @@ def generate_function_definitions(self, env, code):
             if not entry.used or entry.in_closure:
                 continue
 
-            if entry.type.is_memoryviewslice:
-                code.put_xdecref_memoryviewslice(entry.cname, have_gil=not lenv.nogil)
-            elif entry.type.is_pyobject:
-                if not entry.is_arg or len(entry.cf_assignments) > 1:
-                    if entry.xdecref_cleanup:
-                        code.put_var_xdecref(entry)
-                    else:
-                        code.put_var_decref(entry)
+            if entry.type.is_pyobject:
+                if entry.is_arg and not entry.cf_is_reassigned:
+                    continue
+            if entry.xdecref_cleanup:
+                code.put_var_xdecref(entry, have_gil=not lenv.nogil)
+            else:
+                code.put_var_xdecref(entry, have_gil=not lenv.nogil)
```

Comment:
This code screams for either a clear warning comment, or an actual `put_var_decref()` in the second case.

Reply:
Unfortunately it seems to need to be like this. I've created an issue https://github.com/cython/cython/issues/3455 and put a warning comment in for now